### ISSUE TITLE
Multiple "ready" events when watching an array of paths

### ISSUE THREAD

**Issue by spalger** [paulmillr/chokidar#338]
I am currently receiving multiple "ready" events when chokidar is watching less than 200 files, but once the number grows larger I only get one.

I'm using the [following configuration](https://github.com/elastic/kibana/blob/master/src/cli/cluster/ClusterManager.js#L58-L68):
```js
chokidar.watch([
  'src/plugins',
  'src/server',
  'src/ui',
  'src/utils',
  'config',
  'installedPlugins'
], {
  cwd: fromRoot('.'),
  ignored: /[\\\/](node_modules|bower_components|public|__tests__)[\\\/]/,
});
```

**Comment by es128** [comment]
Can you provide a stripped down repo that directly exhibits the problem or more detailed instructions to reproduce using your existing repo?

**Comment by spalger** [comment]
I'll try, but it smells like a race condition and it might be easier for me to just debug chokidar itself. Is there anything specifically that I should keep an eye out for?

**Comment by es128** [comment]
Yes, it does seem like a race condition, and I have a strong suspicion that passing in the array of paths (instead of a single parent dir or glob pattern) is related.

The areas in source to keep an eye on are the treatment of `this._readyCount` and `this._emitReady()`. PR welcome if you identify a fix for the condition you've observed.

Very likely related to #288.

Also, this appears to be a dupe of #286.

**Comment by jdmaurer** [comment]
I was getting this bug when I was watching a file whose parent directory was already being watched.

File structure:
```
node_modules/
styles/
  app.styl
index.js
```

index.js
```
var chokidar = require('chokidar');
chokidar.watch(['styles', 'app.styl'])
  .on('ready', () => {
    console.log('ready'); // <-- this fires twice
  });
```

**Comment by gagern** [comment]
I'm seeing the same. For me, part of the problem appears to be that I'm also trying to watch for files which don't exist yet. For example, inside the chokidar repo,  can use this to reproduce:

```
$ node -e 'c=require("./");c.watch(["noSuchFile","lib"])
    .on("all",function(a,b){console.log(a,b)})
    .on("ready",function(){console.log("=== READY ===")})'
addDir lib
=== READY ===
add lib/fsevents-handler.js
add lib/nodefs-handler.js
=== READY ===
```

One interesting aspect here is that setting `ignoreInitial` suppresses all events, even after the first `ready`. I'm on OS X, in case that matters.

**Comment by gagern** [comment]
Look at [`_emitReady`](https://github.com/paulmillr/chokidar/blob/a5c0b127db9d46617b7a3d82cc3fffc08d2ef164/index.js#L130-L137) calls. We [expect to call this twice per watched item](https://github.com/paulmillr/chokidar/blob/a5c0b127db9d46617b7a3d82cc3fffc08d2ef164/index.js#L608-L609). For an existing object, the first call is [here](https://github.com/paulmillr/chokidar/blob/a5c0b127db9d46617b7a3d82cc3fffc08d2ef164/lib/fsevents-handler.js#L251) and the second [here](https://github.com/paulmillr/chokidar/blob/a5c0b127db9d46617b7a3d82cc3fffc08d2ef164/lib/fsevents-handler.js#L368-L371). In case of an error, however, we have [two successive calls](https://github.com/paulmillr/chokidar/blob/a5c0b127db9d46617b7a3d82cc3fffc08d2ef164/lib/fsevents-handler.js#L330-L331) but still get the one from [here](https://github.com/paulmillr/chokidar/blob/a5c0b127db9d46617b7a3d82cc3fffc08d2ef164/lib/fsevents-handler.js#L251), amounting to three calls. So I take it that one of these is in error, although I'm not sure which one.

@jsonmaur used `['styles', 'app.styl']` but the file is *inside* the `styles` directory, not in the top directory, so it's effectively a missing file as well. So it's the same issue.

**Comment by leoj3n** [comment]
I'm seeing this behavior with latest release. Was this resolved and I'm seeing a regression, Or was this never actually resolved? Would be nice if `ready` didn't fire multiple times.

**Comment by es128** [comment]
@leoj3n if you can set up a reproducible case we can try it with different versions and find out.

**Comment by gagern** [comment]
My fix was merged in fdf8d463a261fd7f25fbca5f8e367a5c38be5185 for #505 but reverted in 08beb85b72a9714cef1d9e2a87ca0fe48aa080ae for #507. So it seems this was at best only fixed for a very short while. I don't have an OS X any more, so I can't check whether my comment above still serves as a reproducting example there.

**Comment by jairmilanes** [comment]
For anyone coming here with the same problem, ready event firing multiple times, I just experience this and what I noticed is if you pass an array of patterns and some of these patterns don't exist, or resolve to no files, the ready event fires multiple times, once I made sure that all patterns existed the ready event begun to fire only once.

Hope it helps! Cheers!
